library/lisparray.lisp

```diff
@@ -0,0 +1,78 @@
+;;;; lisparray.lisp
+;;;;
+;;;; An interface to Common Lisp rank-1 SIMPLE-ARRAYs.
+
+(coalton-library/utils:defstdlib-package #:coalton-library/lisparray
+  (:use #:coalton)
+  (:local-nicknames (#:ty #:coalton-library/types))
+  (:export
+   #:LispArray
+   #:make
+   #:make-uninitialized
+   #:length
+   #:aref
+   #:set!))
+
+(in-package #:coalton-library/lisparray)
+
+(named-readtables:in-readtable coalton:coalton)
+
+#+coalton-release
+(cl:declaim #.coalton-impl/settings:*coalton-optimize-library*)
+
+(coalton-toplevel
+  ;; The representation of (LispArray :t) is specially dealt with by
+  ;; the compiler in lisp-type.lisp.
+  (define-type (LispArray :t)
+    "A one-dimensional, non-resizable array of elements.
+
+These arrays are represented as possibly specialized `(cl:simple-array <type> (cl:*))` and are meant to be used as a tool either to interface with Lisp code or to implement efficient data structures. One should consult `Vector` or `Seq` for more general sequential data structure needs.
+
+Whether or not the arrays are specialized depends on the underlying Lisp implementation. Consult `cl:upgraded-array-element-type` to determine whether `LispArray` may get specialized.")
+
+  (declare make (ty:RuntimeRepr :t => UFix -> :t -> LispArray :t))
+  (define (make n x)
+    "Make a new `LispArray` of length `n` initialized to `x`.
+
+If the type of `x` represents a specialized array "
+    ;; FIXME: how can we get this statically?
+    (let ((type (ty:runtime-repr (ty:proxy-of x))))
+      (lisp (LispArray :t) (n x type)
+        (cl:make-array n :element-type type :initial-element x))))
+
+  (declare make-uninitialized (ty:RuntimeRepr :t => UFix -> ty:Proxy :t -> LispArray :t))
+  (define (make-uninitialized n p)
```

Comment:
You don't need to take a proxy here. [This trick](https://github.com/coalton-lang/coalton/blob/6fdd091375eb52ae6584b20b0165f62bf0dc6678/library/vector.lisp#L73-L76) let's you reference the functions return type.

Reply:
great trick